src/fsharp/FSharp.Compiler.Compilation/AstVisitor.fsi

```diff
@@ -0,0 +1,267 @@
+﻿namespace FSharp.Compiler.Compilation
+
+open System.IO
+open System.Threading
+open System.Collections.Immutable
+open System.Runtime.CompilerServices
+open Microsoft.CodeAnalysis.Text
+open Microsoft.CodeAnalysis.Host
+open FSharp.Compiler.Compilation.Utilities
+open FSharp.Compiler.AbstractIL.Internal.Library
+open FSharp.Compiler
+open FSharp.Compiler.Text
+open FSharp.Compiler.CompileOps
+open FSharp.Compiler.Ast
+open FSharp.Compiler.Range
+
+[<AutoOpen>]
+module AstVisitorHelpers =
+
+    type SynModuleOrNamespace with
+
+        member AdjustedRange: range
+
+    type ParsedInput with
+
+        member AdjustedRange: range
+
+    type ParsedHashDirective with
+
+        member Range: range
+
+    type SynTypeConstraint with
+
+        member Range: range
+
+    type SynMemberSig with
+
+        member Range: range
+
+    type SynValSig with
+
+        member Range: range
+
+    type SynField with
+
+        member Range: range
+
+    type SynTypeDefnSig with
+
+        member Range: range
+
+    type SynMeasure with
+
+        member PossibleRange: range
+
+    type SynRationalConst with
+
+        member PossibleRange: range
+
+    type SynConst with
+
+        member PossibleRange: range
+
+    type SynArgInfo with
+
+        member PossibleRange: range
+
+    type SynValInfo with
+
+        member PossibleRange: range
+
+    type SynTypeDefnKind with
+
+        member PossibleRange: range
+
+    type SynTyparDecl with
+
+        member Range: range
+
+    type SynValTyparDecls with
+
+        member PossibleRange: range
+
+    type SynSimplePat with
+
+        member Range: range
+
+    type SynSimplePats with
+
+        member Range: range
+
+    type SynValData with
+
+        member Range: range
+
+    type SynBindingReturnInfo with
+
+        member Range: range
+
+    type SynConstructorArgs with
+
+        member PossibleRange: range
+
+    type SynInterfaceImpl with
+
+        member Range: range
+
+    type SynSimplePatAlternativeIdInfo with
+
+        member Range: range
+
+    type SynStaticOptimizationConstraint with
+
+        member Range: range
+
+    type SynUnionCaseType with
+
+        member PossibleRange: range
+
+[<AbstractClass>]
+type AstVisitor<'T> =
+
+    new: unit -> AstVisitor<'T>
+
+    abstract CanVisit: range -> bool
```

Comment:
Why it's not a part of the compiler service? I think it would be much better if things that are not Roslyn-specific go to the common code. (Maybe some of the existing service APIs could be deprecated when adding new (better?) ones.

Reply:
This is also to reply above regarding `Range`.

I can imagine these moving back into FCS as they are not really Roslyn-specific. The question is if we want to expose this as a public API. Right now they are just internal implementation so it doesn't really matter where it lives. Just depends on what we decide.